### PR TITLE
Introducing gRPC support

### PR DESCRIPTION
**Pull request by oldergod**

### CODE REVIEW

File: wire-grpc-tests/src/main/java/com/squareup/wire/GrpcWriter.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package com.squareup.wire
```
Comment (swankjesse) [inline wire-grpc-tests/src/main/java/com/squareup/wire/GrpcWriter.kt]: nit: copyright

File: wire-grpc-tests/src/main/java/com/squareup/wire/util.kt [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+internal fun Type.genericParameterType(index: Int = 0): Type {
+  when (this) {
+    is ParameterizedType -> return actualTypeArguments[index]
+    else -> throw IllegalArgumentException("no generic parameter type: $this")
+  }
+}
+
+/**
+ */
+internal fun Type.genericReturnType(): Type {
```
Comment (swankjesse) [inline wire-grpc-tests/src/main/java/com/squareup/wire/util.kt]: Delete?

File: wire-grpc-tests/src/test/java/com/squareup/wire/MockRouteGuideService.kt [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+import org.junit.runners.model.Statement
+import routeguide.RouteGuideGrpc
+import routeguide.RouteGuideProto.Feature
+import routeguide.RouteGuideProto.Point
+import routeguide.RouteGuideProto.Rectangle
+import routeguide.RouteGuideProto.RouteNote
+import routeguide.RouteGuideProto.RouteSummary
+import java.util.ArrayDeque
+
+class MockRouteGuideService : RouteGuideGrpc.RouteGuideImplBase(), TestRule {
```
Comment (swankjesse) [inline wire-grpc-tests/src/test/java/com/squareup/wire/MockRouteGuideService.kt]: needs a couple sentences on how it works!

File: wire-grpc-tests/src/test/java/com/squareup/wire/MockRouteGuideService.kt [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  override fun recordRoute(responseObserver: StreamObserver<RouteSummary>): StreamObserver<Point> {
+    assertThat(script.removeFirst())
+        .isEqualTo(Action.ReceiveCall("/routeguide.RouteGuide/RecordRoute"))
+
+    streamObserver = responseObserver as StreamObserver<Any>
+    processScript()
+
+    return object : StreamObserver<Point> {
```
Comment (swankjesse) [inline wire-grpc-tests/src/test/java/com/squareup/wire/MockRouteGuideService.kt]: extract a function that creates a `StreamObserver<T>` ?

File: wire-grpc-tests/src/test/java/com/squareup/wire/MockRouteGuideService.kt [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    processScript()
+  }
+
+  override fun routeChat(responseObserver: StreamObserver<RouteNote>): StreamObserver<RouteNote> {
+    assertThat(script.removeFirst())
+        .isEqualTo(Action.ReceiveCall("/routeguide.RouteGuide/RouteChat"))
+    streamObserver = responseObserver as StreamObserver<Any>
+    processScript()
+
+    return object : StreamObserver<RouteNote> {
```
Comment (swankjesse) [inline wire-grpc-tests/src/test/java/com/squareup/wire/MockRouteGuideService.kt]: ... and then use it here

File: wire-grpc-tests/src/test/proto-grpc/routeguide/FeatureDatabase.kt [hunk 6]
```diff
@@ -0,0 +1,63 @@
+// Code generated by Wire protocol buffer compiler, do not edit.
+// Source file: routeguide/RouteGuideProto.proto
+package routeguide
```
Comment (swankjesse) [inline wire-grpc-tests/src/test/proto-grpc/routeguide/FeatureDatabase.kt]: can you generate this?

File: wire-tests/src/test/proto-java/routeguide/Feature.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+import java.lang.String;
+import java.lang.StringBuilder;
+import okio.ByteString;
+
+/**
+ * A feature names something at a given point.
+ *
+ * If a feature could not be named, the name is empty.
+ */
+public final class Feature extends Message<Feature, Feature.Builder> {
```
Comment (swankjesse) [inline wire-tests/src/test/proto-java/routeguide/Feature.java]: delete this?

File: wire-grpc-tests/src/main/java/com/squareup/wire/GrpcReader.kt [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  private val messageAdapter: ProtoAdapter<T>,
+  private val grpcEncoding: GrpcEncoding? = null
+) : Closeable {
+  companion object {
+    /**
+     * @param source the HTTP/2 stream body.
+     * @param messageAdapter a proto adapter for each message.
+     * @param grpcEncoding the "grpc-encoding" header, or null if it is absent.
+     */
+    fun <T> get(
```
[the remark this answers is not in the document]
Comment (oldergod) [reply wire-grpc-tests/src/main/java/com/squareup/wire/GrpcReader.kt]: Doesn't really look like so? I removed it and inlined the constructor (by making it public)

File: wire-grpc-tests/src/test/proto-grpc/routeguide/FeatureDatabase.kt (hunk 6 above)
Comment (oldergod) [reply wire-grpc-tests/src/test/proto-grpc/routeguide/FeatureDatabase.kt]: We do generate them but they're all checked in the codebase. Same is happening for current tests.
I'd rather do a follow up to clean the whole project than just us here.

**Pull request merged**